File: project/src/main/java/com/expedia/bookings/widget/CollectionViewHolder.java
package com.expedia.bookings.widget;

import android.os.Bundle;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.View;
import android.widget.ImageView;

import com.expedia.bookings.R;
import com.expedia.bookings.data.collections.CollectionLocation;
import com.expedia.bookings.graphics.HeaderBitmapDrawable;
import com.expedia.bookings.otto.Events;
import com.expedia.bookings.tracking.OmnitureTracking;
import com.expedia.bookings.utils.AnimUtils;
import com.expedia.bookings.utils.FontCache;

import butterknife.ButterKnife;
import butterknife.InjectView;
import butterknife.Optional;

/**
 * A Viewholder for the case where our data are launch collections.
 */
public class CollectionViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, HeaderBitmapDrawable.CallbackListener {
	private static final int FULL_TILE_TEXT_SIZE = 18;
	private static final int HALF_TILE_TEXT_SIZE = 15;

	@Optional
	@InjectView(R.id.card_view)
	public CardView cardView;

	@Optional
	@InjectView(R.id.title)
	public TextView title;

	@Optional
	@InjectView(R.id.subtitle)
	public TextView subtitle;

	@Optional
	@InjectView(R.id.background_image)
	public ImageView backgroundImage;

	@Optional
	@InjectView(R.id.browse_hotels_label)
	public TextView browseHotelsLabel;

	@Optional
	@InjectView(R.id.gradient)
	public View gradient;

	public String collectionUrl;

	public CollectionViewHolder(View view) {
		super(view);
		ButterKnife.inject(this, itemView);
		itemView.setOnClickListener(this);
	}

	public void bindListData(Object data, boolean fullWidthTile, boolean showBrowseHotelsLabel) {
		itemView.setTag(data);
		cardView.setPreventCornerOverlap(false);

		if (fullWidthTile) {
			title.setTextSize(TypedValue.COMPLEX_UNIT_SP, FULL_TILE_TEXT_SIZE);
		}
		else {
			title.setTextSize(TypedValue.COMPLEX_UNIT_SP, HALF_TILE_TEXT_SIZE);
		}
		browseHotelsLabel.setVisibility(showBrowseHotelsLabel ? View.VISIBLE : View.GONE);

		CollectionLocation location = (CollectionLocation) data;
		bindLocationData(location);
	}

	private void bindLocationData(CollectionLocation location) {
		title.setText(location.title);
		FontCache.setTypeface(title, FontCache.Font.ROBOTO_MEDIUM);
		subtitle.setText(location.subtitle);
		subtitle.setVisibility(View.VISIBLE);
	}

	@Override
	public void onClick(View view) {
		Bundle animOptions = AnimUtils.createActivityScaleBundle(view);
		CollectionLocation location = (CollectionLocation) view.getTag();
		Events.post(new Events.LaunchCollectionItemSelected(location, animOptions, collectionUrl));
		OmnitureTracking.trackNewLaunchScreenTileClick(true);
	}

	@Override
	public void onBitmapLoaded() {
		gradient.setVisibility(View.VISIBLE);
	}

	@Override
	public void onBitmapFailed() {
		gradient.setVisibility(View.GONE);
	}

	@Override
	public void onPrepareLoad() {
		gradient.setVisibility(View.GONE);
	}
}

File: project/src/main/java/com/expedia/bookings/widget/CollectionViewHolder.kt
package com.expedia.bookings.widget

import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.View
import android.widget.ImageView
import com.expedia.bookings.R
import com.expedia.bookings.data.collections.CollectionLocation
import com.expedia.bookings.graphics.HeaderBitmapDrawable
import com.expedia.bookings.otto.Events
import com.expedia.bookings.tracking.OmnitureTracking
import com.expedia.bookings.utils.AnimUtils
import com.expedia.bookings.utils.FontCache
import com.expedia.bookings.utils.bindView

/**
 * A Viewholder for the case where our data are launch collections.
 */
class CollectionViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener, HeaderBitmapDrawable.PicassoTargetListener {

    val cardView: CardView by bindView (R.id.card_view)
    val title: TextView by bindView(R.id.title)
    val subtitle: TextView by bindView(R.id.subtitle)
    val backgroundImage: ImageView by bindView(R.id.background_image)
    val browseHotelsLabel: TextView by bindView(R.id.browse_hotels_label)
    val gradient: View by bindView(R.id.gradient)
    var collectionUrl = ""

    init {
        itemView.setOnClickListener(this)
    }

    fun bindListData(data: Any, fullWidthTile: Boolean, showBrowseHotelsLabel: Boolean) {
        itemView.tag = data
        cardView.preventCornerOverlap = false

        if (fullWidthTile) {
            title.setTextSize(TypedValue.COMPLEX_UNIT_SP, FULL_TILE_TEXT_SIZE.toFloat())
        } else {
            title.setTextSize(TypedValue.COMPLEX_UNIT_SP, HALF_TILE_TEXT_SIZE.toFloat())
        }
        browseHotelsLabel.visibility = if (showBrowseHotelsLabel) View.VISIBLE else View.GONE

        val location = data as CollectionLocation
        bindLocationData(location)
    }

    private fun bindLocationData(location: CollectionLocation) {
        title.text = location.title
        FontCache.setTypeface(title, FontCache.Font.ROBOTO_MEDIUM)
        subtitle.text = location.subtitle
        subtitle.visibility = View.VISIBLE
    }

    override fun onClick(view: View) {
        val animOptions = AnimUtils.createActivityScaleBundle(view)
        val collection = view.tag as CollectionLocation
        Events.post(Events.LaunchCollectionItemSelected(collection, animOptions, collectionUrl))
        OmnitureTracking.trackNewLaunchScreenTileClick(true)
    }

    override fun onBitmapLoaded() {
        gradient.visibility = View.VISIBLE
    }

    override fun onBitmapFailed() {
        gradient.visibility = View.GONE
    }

    override fun onPrepareLoad() {
        gradient.visibility = View.GONE
    }

    companion object {
        private val FULL_TILE_TEXT_SIZE = 18
        private val HALF_TILE_TEXT_SIZE = 15
    }
}
